crates/ironrdp-cliprdr/src/backend/windows.rs

```diff
@@ -0,0 +1,165 @@
+use self::cliprdr_backend::WinCliprdrBackend;
+
+use crate::backend::{ClipboardMessageProxy, CliprdrBackendFactory};
+use crate::pdu::{
+    ClipboardFormat, ClipboardFormatId, ClipboardGeneralCapabilityFlags, FormatDataRequest, FormatDataResponse,
+};
+use crate::CliprdrBackend;
+
+use thiserror::Error;
+use tracing::error;
+use winapi::shared::basetsd::DWORD_PTR;
+use winapi::shared::windef::HWND;
+use winapi::um::commctrl::{RemoveWindowSubclass, SetWindowSubclass};
+use winapi::um::winuser::{AddClipboardFormatListener, RemoveClipboardFormatListener, WM_USER};
+
+use std::sync::{mpsc as mpsc_sync, Arc};
+
+const BACKEND_CHANNEL_SIZE: usize = 8;
+const WM_USER_CLIPBOARD: u32 = WM_USER;
+
+pub type WinCliprdrResult<T> = Result<T, WinCliprdrError>;
+
+#[derive(Debug, Error)]
+pub enum WinCliprdrError {
+    #[error("Failed to register clipboard format listener")]
+    AddClipboardFormatListener,
+
+    #[error("Failed to enumerate formats available in the current clipboard")]
+    FormatsEnumeration,
+
+    #[error("Clipboard is busy")]
+    ClipboardAccessDenied,
+
+    #[error("Failed to open the clipboard")]
+    ClipboardOpen,
+
+    #[error("Failed to empty the clipboard")]
+    ClipboardEmpty,
+
+    #[error("Failed to convert UTF-16 string to UTF-8")]
+    Uft16Conversion,
+
+    #[error("Failed to get current clipboard data")]
+    ClipboardData,
+
+    #[error("Failed to set clipboard data")]
+    SetClipboardData,
+
+    #[error("Failed to subclass window")]
+    WindowSubclass,
+
+    #[error("Failed to allocate global memory")]
+    Alloc,
+
+    #[error("Failed to receive data from remote clipboard")]
+    DataReceiveTimeout,
+
+    #[error("Failed to render clipboard format")]
+    RenderFormat,
+}
+
+/// Sent from the clipboard backend shim to the actual WinAPI subproc event loop
+#[derive(Debug)]
+pub(crate) enum BackendEvent {
+    // Events generated by OS event loop
+    ClipboardUpdated,
+    RenderFormat(ClipboardFormatId),
+    RenderAllFormats,
+
+    // PDU processing events
+    DowngradedCapabilities(ClipboardGeneralCapabilityFlags),
+    RemoteFormatList(Vec<ClipboardFormat>),
+    FormatDataRequest(FormatDataRequest),
+    FormatDataResponse(FormatDataResponse<'static>),
+    RemoteRequestsFormatList,
+}
+
+/// Windows RDP client clipboard implementation.
+///
+/// IronRDP client implementation should provide raw window handle and message proxy to send
+/// messages from the backend to `CLIPRDR` SVC.
+pub struct WinClipboard {
+    /// HWND is declated as non-Send type, but it is actually safe to send and use it between threads,
+    /// as long as GUI updates are not performed from the background thread. We use workaround here
+    /// to keep the handle as usize instead of pointer
+    window: usize,
+    backend_tx: mpsc_sync::SyncSender<BackendEvent>,
+}
+
+impl WinClipboard {
+    /// Creates new clipboard instance.
+    #[allow(clippy::not_unsafe_ptr_arg_deref)] // False positive, HWND is a handle, not a pointer
```

Comment:
:red_square: It is essentially a pointer and it’s not guaranteed that it’s safe to pass a garbage handle AFAIK, and this function is basically forwarding the handle to an unsafe function. Note that this handle also has a validity lifetime that is elided. This function should be marked as unsafe, and the documentation should specify the caller must "ensure that the provided object is valid for the lifetime of the WinClipboard instance".

Also, there might be some value to integrate with [the `raw_window_handle` crate](https://lib.rs/crates/raw-window-handle) here. I’m not sure though, since this is already platform-specific code in this case.

You could check softbuffer’s Context for reference: https://docs.rs/softbuffer/latest/softbuffer/struct.Context.html#method.new

Reply:
Agreed, thanks for pointing it out, I'll mark it unsafe. Now I am thinking about that again, it is definitely was not a false positive indeed!